cirq/contrib/paulistring/recombine.py

```diff
@@ -50,13 +52,13 @@ def _possible_string_placements(
                 break
             string_op = string_op.pass_operations_over([out_op],
```

Comment:
Please add a test that exercises this line when `out_op` is another pauli string placed earlier (this is the corner case I described).  This case didn't happen before this PR.

Reply:
thank you for the analysis, this is cool, your derivation seems sound to me. 

And this argument holds even after merging PauliStrings with other PauliStrings that equal to it up to a coefficient - as that doesn't change the commutativity relations. 

I'm not sure I can cook up an interesting test case for this: 

Rephrasing your situation: 

- in the DAG the Paulistring A' has an ancestor B' that it doesn't commute with 
- A' and C' are in the frontier (commuting) 
- C' commutes with B as well

```
B' --- A'
       C'
```

In order to get the desired "conflict", we'll need to have: 
- A' has to be further than C' - so that it can be picked first from the frontier 
- B' has to be further in the circuit than C' - so the "conflict" arises

Thus: 
- A' gets inserted first 
- then C' is inserted 
- then B' needs to commute over C' to get to its place 

The only way the original single qubit paulistrings will commute is if they are equal up to a coefficient or they are on a different qubit. Having all of them in the same qubit leads to contradiction: C ~= A ~= B => [C,B]=0 => [C',B'] = 0 that contradicts our initial setup. Hence they have to be on separate qubits. 

```
...-----@-- B ------ A ----- 
...     |
...-C---@-----------------
```

To cover the line you want to cover with op being a paulistring, we need to make sure that as B' is passing through, it is a multiqubit PauliString that has a Pauli D' on the qubit where C lives and D' ~= C. Otherwise if the qubits are disjoint, we just simply skip the passing. 

Thinking...

Reply:
It was very hard to find a circuit to do this but I think I found one that does.  I only checked it by hand so I may have made a mistake.  The key was making [3] operate on another qubit that [2] didn't operate on and making [1] operate on another qubit that [2] and [3] don't operate on.

```
q0, q1, q2 = cirq.LineQubit.range(3)
strings = cirq.Circuit.from_ops(
    (cirq.Z(q1)*cirq.X(q2))**0.3,  # Operation [3]
    (cirq.X(q2))**0.2,  # Operation [2]
    (cirq.Z(q0)*cirq.X(q1))**0.1,  # Operation [1]
)
circuit = cirq.Circuit.from_ops(
    cirq.CNOT(q2, q1),
    cirq.SingleQubitCliffordGate.Z_sqrt(q2),
    cirq.SingleQubitCliffordGate.X_sqrt(q1),
    cirq.SingleQubitCliffordGate.X_sqrt(q1),
    cirq.SingleQubitCliffordGate.Y_sqrt(q0),
    cirq.CZ(q0, q1),
    cirq.SingleQubitCliffordGate.X_sqrt(q1),
    cirq.SingleQubitCliffordGate.Y_sqrt(q0),
    cirq.CZ(q1, q2),
    cirq.CX(q1, q0),
    cirq.SingleQubitCliffordGate.X_sqrt(q0),
    cirq.CNOT(q0, q2),
    cirq.SingleQubitCliffordGate.X_sqrt(q0),
    strategy=cirq.InsertStrategy.NEW_THEN_INLINE,
)
strings + circuit
```
```
0: ─────────────[Z]───────    ───────────────Y^0.5───@───Y^0.5───────X───X^0.5───@───X^0.5───
                │                                    │               │           │
1: ───[Z]───────[X]^0.1───    ───X───X^0.5───X^0.5───@───X^0.5───@───@───────────┼───────────
      │                          │                               │               │
2: ───[X]^0.3───X^0.2─────    ───@───Z^0.5───────────────────────@───────────────X───────────
```

- [1] is placed at the far end of the circuit on qubit 0.
- [2] is can only be placed on the left before the first CNOT.
- [3] commutes past [2] and becomes a single qubit pauli string right before the last X^0.5.

The circuit could probably be simplified.

Reply:
I just had time to quickly try this out, will look into this deeper in the days. 

After running `dag = pauli_string_dag_from_circuit(c)` tried both the old and the new insertion algo, the new ops seem to get stuck at the front of the circuit: 

![image](https://user-images.githubusercontent.com/635073/61592557-8f074b80-ab89-11e9-9e6f-19a00ca0aa56.png)

Reply:
I think it's because CNOT is not part of the gateset, CZ is.

Reply:
Yes.  You can replace `CNOT(a, b)` with `[cirq.SingleQubitCliffordGate.Y_sqrt(b), CZ(a, b), cirq.SingleQubitCliffordGate.Y_nsqrt(b)]`.  (I may have the signs of the Y gates flipped.)

Reply:
Ah yes, but they commute...which we check for before. i.e. the desired line won't be executed.
I commented with a simpler case that achieves the same, plus here's my derivation for the partial circuit matching - commutativity still holds even when we move the ops just part of the circuit: 

![image](https://user-images.githubusercontent.com/635073/61661487-053aa980-ac81-11e9-8006-b906712d8e54.png)

Reply:
I didn't see that check.  I guess the code already handles this case.  That's what I wanted to trigger with the test.

I'll approve this once you address the other comment.

Reply:
Done!

Reply:
Thank you again for the thorough review!